user/store.go

```diff
@@ -227,18 +253,143 @@ func (db *DB) Insert(ctx context.Context, u *User) (*User, error) {
 	return u, nil
 }
 
+func (db *DB) DeleteTx(ctx context.Context, tx *sql.Tx, id string) error {
+	err := permission.LimitCheckAny(ctx, permission.System, permission.Admin)
+	if err != nil {
+		return err
+	}
+
+	err = validate.UUID("UserID", id)
+	if err != nil {
+		return err
+	}
+	return db.retryDeleteTx(ctx, tx, id)
+}
+
+func (db *DB) retryDeleteTx(ctx context.Context, tx *sql.Tx, id string) error {
+	return retry.DoTemporaryError(func(int) error {
+		err := db._deleteTx(ctx, tx, id)
+		sqlErr := sqlutil.MapError(err)
+		if sqlErr != nil && sqlErr.Code == "23503" {
+			// retry foreign key errors when deleting a user
+			err = retry.TemporaryError(err)
+		}
+		return err
+	},
+		retry.Log(ctx),
+		retry.Limit(5),
+		retry.FibBackoff(250*time.Millisecond),
+	)
+}
+
+func (db *DB) _deleteTx(ctx context.Context, tx *sql.Tx, id string) error {
+	// cleanup rotations first
+	rows, err := tx.StmtContext(ctx, db.userRotations).QueryContext(ctx, id)
+	if errors.Is(err, sql.ErrNoRows) {
+		err = nil
+	}
+	if err != nil {
+		return fmt.Errorf("lookup user rotations: %w", err)
+	}
+	defer rows.Close()
+
+	var rotationIDs []string
+	for rows.Next() {
+		var rID string
+		err = rows.Scan(&rID)
+		if err != nil {
+			return fmt.Errorf("scan user rotation id: %w", err)
+		}
+		rotationIDs = append(rotationIDs, rID)
+	}
+
+	for _, rID := range rotationIDs {
+		err = db.removeUserFromRotation(ctx, tx, id, rID)
+		if err != nil {
+			return fmt.Errorf("remove user '%s' from rotation '%s': %w", id, rID, err)
+		}
+	}
+
+	_, err = tx.StmtContext(ctx, db.deleteOne).ExecContext(ctx, id)
+	if err != nil {
+		return fmt.Errorf("delete user row: %w", err)
+	}
+	return nil
+}
+
+func (db *DB) removeUserFromRotation(ctx context.Context, tx *sql.Tx, userID, rotationID string) error {
+	type part struct {
+		ID     string
+		UserID string
+	}
+	var participants []part
+	rows, err := tx.StmtContext(ctx, db.rotationParts).QueryContext(ctx, rotationID)
+	if err != nil {
+		return fmt.Errorf("query participants: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var p part
+		err = rows.Scan(&p.ID, &p.UserID)
+		if err != nil {
+			return fmt.Errorf("scan participant %d: %w", len(participants), err)
```

Comment:
It's to log the index/position number since we're processing them in order, tells us if it failed on the 1st vs 3rd scan. Since the scan failed, we don't have `p.ID` and it would just be empty.